pkg/types/tuf/v0.0.1/entry.go

```diff
@@ -0,0 +1,251 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package tuf
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	//"io"
+	//"io/ioutil"
+	"reflect"
+	//"strconv"
+	"strings"
+
+	"github.com/sigstore/rekor/pkg/log"
+	"github.com/sigstore/rekor/pkg/types"
+	"github.com/sigstore/rekor/pkg/types/tuf"
+	"github.com/sigstore/rekor/pkg/util"
+
+	"github.com/go-openapi/strfmt"
+
+	"github.com/sigstore/rekor/pkg/pki"
+
+	"github.com/go-openapi/swag"
+	"github.com/mitchellh/mapstructure"
+	"github.com/sigstore/rekor/pkg/generated/models"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	APIVERSION = "0.0.1"
+)
+
+func init() {
+	tuf.SemVerToFacFnMap.Set(APIVERSION, NewEntry)
+}
+
+type V001Entry struct {
+	TufModel                models.TufV001Schema
+	fetchedExternalEntities bool
+	keyObj                  pki.PublicKey
+}
+
+func (v V001Entry) APIVersion() string {
+	return APIVERSION
+}
+
+func NewEntry() types.EntryImpl {
+	return &V001Entry{}
+}
+
+func Base64StringtoByteArray() mapstructure.DecodeHookFunc {
+	return func(f reflect.Type, t reflect.Type, data interface{}) (interface{}, error) {
+		if f.Kind() != reflect.String || t.Kind() != reflect.Slice {
+			return data, nil
+		}
+
+		bytes, err := base64.StdEncoding.DecodeString(data.(string))
+		if err != nil {
+			return []byte{}, fmt.Errorf("failed parsing base64 data: %v", err)
+		}
+		return bytes, nil
+	}
+}
+
+func (v V001Entry) IndexKeys() []string {
+	var result []string
+
+	if v.HasExternalEntities() {
+		if err := v.FetchExternalEntities(context.Background()); err != nil {
+			log.Logger.Error(err)
+			return result
+		}
+	}
+
+	key, err := v.keyObj.CanonicalValue()
+	if err != nil {
+		log.Logger.Error(err)
+	} else {
+		hasher := sha256.New()
+		if _, err := hasher.Write(key); err != nil {
+			log.Logger.Error(err)
+		} else {
+			result = append(result, strings.ToLower(hex.EncodeToString(hasher.Sum(nil))))
+		}
+	}
+
+	return result
+}
+
+func (v *V001Entry) Unmarshal(pe models.ProposedEntry) error {
+	tuf, ok := pe.(*models.Tuf)
+	if !ok {
+		return errors.New("cannot unmarshal non tuf v0.0.1 type")
+	}
+
+	cfg := mapstructure.DecoderConfig{
+		DecodeHook: Base64StringtoByteArray(),
+		Result:     &v.TufModel,
+	}
+
+	dec, err := mapstructure.NewDecoder(&cfg)
+	if err != nil {
+		return fmt.Errorf("error initializing decoder: %w", err)
+	}
+
+	if err := dec.Decode(tuf.Spec); err != nil {
+		return err
+	}
+	// field validation
+	if err := v.TufModel.Validate(strfmt.Default); err != nil {
+		return err
+	}
+	// cross field validation
+	return v.Validate()
+
+}
+
+func (v V001Entry) HasExternalEntities() bool {
+	if v.fetchedExternalEntities {
+		return false
+	}
+
+	if v.TufModel.PublicKey != nil && v.TufModel.PublicKey.URL.String() != "" {
+		return true
+	}
+
+	return false
+}
+
+func (v *V001Entry) FetchExternalEntities(ctx context.Context) error {
+
```

Comment:
should check to see if `v.fetchedExternalEntities == true` here and bail early if it has